src/dataflow-types/src/reconciliation/command.rs

```diff
@@ -0,0 +1,361 @@
+    /// Did we receive a STORAGE CreateInstance command?
+    storage_initialized: bool,
+    /// Upper frontiers for indexes, sources, and sinks.
+    ///
+    /// The `Option<ComputeInstanceId>` uses `None` to represent Storage.
+    uppers: HashMap<(GlobalId, Option<ComputeInstanceId>), MutableAntichain<T>>,
+}
+
+#[async_trait(?Send)]
+impl<C: Client + 'static> Client for CommandReconcile<mz_repr::Timestamp, C> {
+    async fn send(&mut self, cmd: Command) -> Result<(), Error> {
+        self.observe_command(cmd).await
+    }
+
+    async fn recv(&mut self) -> Option<Response> {
+        if let Some(response) = self.responses.pop_front() {
+            Some(response)
+        } else {
+            let response = self.client.recv().await;
+            if let Some(response) = response {
+                self.absorb_response(response)
+            }
+            self.responses.pop_front()
+        }
+    }
+}
+
+impl<C: Client> CommandReconcile<mz_repr::Timestamp, C> {
+    /// Construct a new [CommandReconcile].
+    ///
+    /// * `client`: The client wrapped by this struct.
+    pub fn new(client: C) -> Self {
+        Self {
+            client,
+            created: false,
+            responses: Default::default(),
+            storage_initialized: false,
+            uppers: Default::default(),
+        }
+    }
+
+    /// Start tracking of a id within an instance.
+    ///
+    /// If we're already tracking this ID, it means that the coordinator lost connection and
+    /// reconnected (or has a bug). We're updating the coordinator's upper to match the local state.
+    fn start_tracking(&mut self, id: GlobalId, instance: Option<ComputeInstanceId>) {
+        let frontier = timely::progress::frontier::MutableAntichain::new_bottom(
+            <mz_repr::Timestamp as timely::progress::Timestamp>::minimum(),
+        );
+        match self.uppers.entry((id, instance)) {
+            Entry::Occupied(entry) => {
+                let mut change_batch = ChangeBatch::new_from(
+                    <mz_repr::Timestamp as timely::progress::Timestamp>::minimum(),
+                    -1,
+                );
+                change_batch.extend(entry.get().frontier().iter().copied().map(|t| (t, 1)));
+                // dbg!(entry.get());
+                // dbg!(&change_batch);
+                if let Some(instance) = instance {
+                    self.responses.push_back(Response::Compute(
+                        ComputeResponse::FrontierUppers(vec![(id, change_batch)]),
+                        instance,
+                    ));
+                } else {
+                    let feedback = TimestampBindingFeedback {
+                        changes: vec![(id, change_batch)],
+                        bindings: vec![],
+                    };
+                    self.responses.push_back(Response::Storage(
+                        StorageResponse::TimestampBindings(feedback),
+                    ));
+                }
```

Comment:
This stuff is a bit gross and I hope we fix it. :D

I think technically this is a potential definiteness error, because the intended meaning of `TimestampBindingFeedback` is "this frontier is definite only once you write down these bindings". So, a "correct" implementation would probably retain the bindings that need to be written down (up through what compaction has been allowed). Big pain in the butt, and ideally STORAGE takes ownership of all of this and it is less painful here at least.

Reply:
I concur that this is gross. I'm leaning towards leaving it as-is for the moment and ask someone from STORAGE to take ownership of this part later. That'll keep my mental load lower at the moment, because I don't know what the semantics of the `TimestampBindings` are :)